### PR TITLE
New Select: Initial scaffolding

### LINKED ISSUE #88897: New Select: Create base scaffolding for the component
- Should render, unstyled, text input and options
 - Should have props with a good guess of what we think the API should be

### PR DESCRIPTION
**Pull request by tskarhed**
**What is this feature?**

Adds a new unstyled Combobox component that uses Downshift

https://github.com/grafana/grafana/assets/1438972/82cc1f24-6f4d-45da-954f-36b2493608e0

**Why do we need this feature?**

Reduce complexity and maintenance of original Select

**Who is this feature for?**

Developers

**Which issue(s) does this PR fix?**:

Closes [#88897](https://github.com/grafana/grafana/issues/88897)

**Special notes for your reviewer:**

- Is there a way we can enforce label use? It is a feature we can use from Downshift, but it all depends on which context this component is used

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { useCombobox } from 'downshift';
+import React, { useState } from 'react';
+
+import { SelectableValue } from '@grafana/data';
+
+import { Input } from '../Input/Input';
+
+interface SimpleSelectProps {
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx:8]: These could probably extend `Input`? At the very least we should pass down invalid and loading to Input.

File: packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  return (item: SelectableValue) => {
+    return (
+      !inputValue ||
+      item?.label?.toLowerCase().includes(lowerCasedInputValue) ||
+      item?.value?.toLowerCase().includes(lowerCasedInputValue)
+    );
+  };
+}
+
+export const SimpleSelect = ({ options, onChange, value, ...restProps }: SimpleSelectProps) => {
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx:32]: I know we're still very early on - lets call this Combobox for now.

File: packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import { useCombobox } from 'downshift';
+import React, { useState } from 'react';
+
+import { SelectableValue } from '@grafana/data';
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx:4]: SelectableValue is very wide, and has a bunch of extra properties that we're not going to use, at least to begin with.

I think we should make a new interface here, at least to begin with. Maybe something like

```ts
type Value= string | number;
type Option = {
  label: string;
  value: Value;
}
```

and then we go from there

File: packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import { Icon } from '../Icon/Icon';
+import { Input, Props as InputProps } from '../Input/Input';
+
+interface SimpleSelectProps
+  extends Omit<InputProps, 'width' | 'prefix' | 'suffix' | 'value' | 'addonBefore' | 'addonAfter'> {
+  onChange: (val: SelectableValue) => void;
+  value: SelectableValue;
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx:12]: My gut feeling is that `value` should be a primative (`string` or `string | number`), to make it easier to pick the correct value. I think making it be an object begins an awkward slipery slope of

```ts
<SimpleSelect
  value={{ label: "Hello!", value: "hello" }}
  options={[
    { label: "Hello!", value: "hello" },
    { label: "World!", value: "world" },
  ]}
/>
```

Which value, if any, is currently selected? 🤔 

```ts
<SimpleSelect
  value={{ label: "foo!", value: "hello" }}
  options={[
    { label: "Hello!", value: "hello" },
    { label: "World!", value: "world" },
  ]}
/>
```

What about now?


I think restricting `value` to just `string`, we remove a bunch of ambiguity.

**Comment by leeoniya** [comment]
please test virtualization performance with this approach (e.g. 10k-100k items) :smile: 

i made this demo [1] a while back for live-filtering a 20k virtualized list. would be nice to get close to this.

[1] https://codesandbox.io/s/musing-mccarthy-pwx88g?file=/index.js

**Review by joshhunt — COMMENTED** [review]
Nice stuff! This is a great start. The main thing think should be changed is to not use SelectableValue just yet.

File: packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx (hunk 3 above)
Comment (tskarhed) [reply packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx]: True, looking at it, it seems like it was introduced specifically for Select. I think the `[key: string]: any;` kinda defeats its purpose a bit. Let's use this opportunity to lock things down 😄

File: packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx (hunk 4 above)
Comment (tskarhed) [reply packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx]: I agree it makes sense. I think with a good example it will be hard to miss too.

**Comment by tskarhed** [comment]
> please test virtualization performance with this approach (e.g. 10k-100k items) 😄
> 
> i made this demo [1] a while back for live-filtering a 20k virtualized list. would be nice to get close to this.
> 
> [1] https://codesandbox.io/s/musing-mccarthy-pwx88g?file=/index.js

Thanks, we'll have a look. That is literally the next step!

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  };
+}
+
+export const Combobox = ({ options, onChange, value, ...restProps }: ComboboxProps) => {
+  const [items, setItems] = useState(options);
+
+  const { getInputProps, getMenuProps, getItemProps, isOpen } = useCombobox({
+    items,
+    itemToString,
+    selectedItem: options.find((option) => option.value === value) || null,
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.tsx:44]: @joshhunt This is the side effect of not passing the object as a value

Comment (joshhunt) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:44]: Nice. Lets memoize it! 👍

File: packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx (hunk 3 above)
Comment (dprokop) [reply packages/grafana-ui/src/components/SimpleSelect/SimpleSelect.tsx]: +1 to @joshhunt comment. If we have a chance to rethink the interface, let's do it. However, keep in mind that `SelectableValue` holds certain properties for a reason (i.e. `description` or `imageUrl`) that may be useful if we are aiming at replacing some usages of the Select component.

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx (hunk 5 above)
Comment (joshhunt) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:44]: Nice! Lets `useMemo` it as well to make sure we don't loop over more than we need to!

**Comment by dprokop** [comment]
> please test virtualization performance with this approach (e.g. 10k-100k items) 😄

Would it make sense to have the new Select virtualised by default? I truly dislike the current Select having to be explicitly configured to be virtualised.

**Comment by joshhunt** [comment]
As I hope is apparent, this component is far from complete! This PR is a starting point for all the other work that must be done in order to make this shippable and an improvement on what we have now. You can keep an eye on the progress with the epic https://github.com/grafana/grafana/issues/87959

File: packages/grafana-ui/src/components/Combobox/Combobox.tsx (hunk 5 above)
Comment (tskarhed) [reply packages/grafana-ui/src/components/Combobox/Combobox.tsx:44]: @dprokop The multi select will initially be its own thing, then we'll look at if we can extract anything.

**Review by joshhunt — APPROVED** [review]
Looks _great_ to me! I think this is a good start to iterate on!

**Pull request merged**